### PR TITLE
[Resources.Gcp] Add Google Cloud Platform resource detector

### PR DESCRIPTION
**Pull request by matt-hensley**
## Changes

Adds Google Cloud Platform resource detector. See package README for published resource attributes.

This package use the new namespace and extension method approach proposed in #1610.

A `AddGcpDetector` extension method is available for `ResourceBuilder`. 

* [x] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #
* [x] Changes in public API reviewed

### CODE REVIEW

File: src/OpenTelemetry.ResourceDetectors.Gcp/GcpResourceDetector.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using System.Collections.Generic;
+using Google.Api.Gax;
+using OpenTelemetry.Resources;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.ResourceDetectors.Gcp;
```
Comment (cijothomas) [inline src/OpenTelemetry.ResourceDetectors.Gcp/GcpResourceDetector.cs:9]: Would it be better if we start this one with the proposal from https://github.com/open-telemetry/opentelemetry-dotnet-contrib/issues/1610?

**Comment by rajkumar-rangaraj** [comment]
This is a new component and it requires a code owner. If you are planning to own this component, could you please update the following file: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/.github/component_owners.yml?

File: src/OpenTelemetry.Resources.Gcp/GcpResourceDetector.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+using System.Collections.Generic;
+using Google.Api.Gax;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Resources.Gcp;
+
+/// <summary>
+/// Resource detector for Google Cloud Platform (GCP).
+/// </summary>
+public sealed class GcpResourceDetector : IResourceDetector
```
Comment (cijothomas) [inline src/OpenTelemetry.Resources.Gcp/GcpResourceDetector.cs:13]: nit: you can keep this internal. The extension method is public, and that should be good enough for users.

**Review by cijothomas — APPROVED** [review]
Not a GCP expert, but looks good from OpenTelemetry standpoint.

File: .github/workflows/package-ResourceDetectors.Gcp.yml [hunk 3]
```diff
@@ -0,0 +1,21 @@
+name: Pack OpenTelemetry.Resources.Gcp
```
Comment (cijothomas) [inline .github/workflows/package-ResourceDetectors.Gcp.yml:1]: The name of this file can be renamed to remove "Detector" part.

**Review by rajkumar-rangaraj — APPROVED** [review]

File: src/OpenTelemetry.Resources.Gcp/OpenTelemetry.Resources.Gcp.csproj [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    <PropertyGroup>
+        <TargetFrameworks>net6.0;netstandard2.0</TargetFrameworks>
+        <Description>OpenTelemetry Resource Detectors for Google Cloud Platform environments</Description>
+        <PackageTags>$(PackageTags);ResourceDetector</PackageTags>
+        <MinVerTagPrefix>Resources.Gcp-</MinVerTagPrefix>
+    </PropertyGroup>
+
+    <ItemGroup>
+        <PackageReference Include="Google.Api.Gax" Version="4.8.0" />
+        <PackageReference Include="Grpc.Core.Api" Version="2.62.0" />
```
Comment (vishweshbankwar) [inline src/OpenTelemetry.Resources.Gcp/OpenTelemetry.Resources.Gcp.csproj:11]: is this needed?

File: src/OpenTelemetry.Resources.Gcp/GcpResourceDetector.cs (hunk 2 above)
Comment (matt-hensley) [reply src/OpenTelemetry.Resources.Gcp/GcpResourceDetector.cs]: Wasn't sure if that part of #1610 was going to be adopted 👍

File: .github/component_owners.yml [hunk 5]
```diff
@@ -79,6 +79,8 @@ components:
   src/OpenTelemetry.ResourceDetectors.ProcessRuntime/:
     - Kielek
     - lachmatt
+  src/OpenTelemetry.Resources.Gcp/:
```
Comment (Kielek) [inline .github/component_owners.yml:82]: Please add similar entry for tests project.

File: .github/component_owners.yml [hunk 6]
```diff
@@ -79,6 +79,8 @@ components:
...
+    - matt-hensley
```
Comment (Kielek) [inline .github/component_owners.yml:83]: It is good practice to have at least 2 component owners. Do you have any other OTel contributor in your organization who can work as the second owner?

File: .github/workflows/package-Resources.Gcp.yml
Comment (Kielek) [inline .github/workflows/package-Resources.Gcp.yml:1]: Need one more file under ISSUE_TEMPLATE folder. See this commit as example: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1694/commits/f6f6ecffafaa580dfd7699405cebf447dc64b3bd

Please use this label: `comp:resources.gcp`

File: src/OpenTelemetry.Resources.Gcp/CHANGELOG.md [hunk 7]
```diff
@@ -0,0 +1,9 @@
+# Changelog
+
+## Unreleased
+
+* Add Google Cloud Platform resource detector for GKE, GAE, GCR, and GCE. Dectector
+is accessible via `AddGcpDetector` extension method on `ResourceBuilder`.
+ ([#1691](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1691))
```
Comment (Kielek) [inline src/OpenTelemetry.Resources.Gcp/CHANGELOG.md:7]:
```suggestion
* Add Google Cloud Platform resource detector for GKE, GAE, GCR, and GCE. Dectector
  is accessible via `AddGcpDetector` extension method on `ResourceBuilder`.
  ([#1691](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/1691))
```

**Review by Kielek — APPROVED** [review]
LGTM.

@CodeBlanch, could you please review public API?

File: src/OpenTelemetry.Resources.Gcp/.publicApi/net6.0/PublicAPI.Shipped.txt
Comment (CodeBlanch) [inline src/OpenTelemetry.Resources.Gcp/.publicApi/net6.0/PublicAPI.Shipped.txt:1]: @matt-hensley 

FYI you can fold these `net6.0` and `netstandard2.0` `.publicApi` files into a single set if you want. Save yourself a bit of maintenance 😄

Just have `.publicApi\PublicAPI.Shipped.txt` & `.publicApi\PublicAPI.Unshipped.txt`.

This is a newer thing in this repo but we have been doing it for a while in the main repo.

See: #1671 & #1675

File: src/OpenTelemetry.Resources.Gcp/GcpResourceBuilderExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+using OpenTelemetry.Internal;
+using OpenTelemetry.Resources.Gcp;
+
+namespace OpenTelemetry.Resources;
```
Comment (CodeBlanch) [inline src/OpenTelemetry.Resources.Gcp/GcpResourceBuilderExtensions.cs:7]: Should extensions go into the namespace of the thing they are extending or somewhere else?

We have had debates about this in the past so I thought I would drop a comment here to make it known this was an intentional decision.

[.NET guidelines](https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/extension-methods) say this:

image

I would argue these packages/extensions are all about dependency management for users so this gets a 👍 from me.

**Comment by CodeBlanch** [comment]
The public API LGTM. This whole PR is very nicely done! I ❤️ using `OpenTelemetry.Resources` in the namespace and the package name.

I haven't approved just yet because I asked @alanwest to also take a look. Because he was originally leading the charge for better package names (https://github.com/open-telemetry/opentelemetry-dotnet/issues/3469).

**Comment by alanwest** [comment]
> I ❤️ using OpenTelemetry.Resources in the namespace and the package name.

I don't mind if we change course on this, but is the plan to rename the [existing packages](https://www.nuget.org/packages?q=opentelemetry.resourcedetectors) to `OpenTelemetry.Resources.*`?

**Comment by Kielek** [comment]
> > I ❤️ using OpenTelemetry.Resources in the namespace and the package name.
> 
> I don't mind if we change course on this, but is the plan to rename the [existing packages](https://www.nuget.org/packages?q=opentelemetry.resourcedetectors) to `OpenTelemetry.Resources.*`?

Yes. #1610

File: src/OpenTelemetry.Resources.Gcp/GcpResourceBuilderExtensions.cs (hunk 8 above)
Comment (matt-hensley) [reply src/OpenTelemetry.Resources.Gcp/GcpResourceBuilderExtensions.cs:7]: Most of the instrumentations put their extension methods in the namespace for the type they're extending (ex. `OpenTelemetry.Trace`, `OpenTelemetry.Metrics`), so would fit unless the other extension methods are going to be moved.

**Review by reyang — APPROVED** [review]

**Review by CodeBlanch — APPROVED** [review]
LGTM

**Pull request merged**